### ISSUE TITLE
Competition with other sources of virtual text

### ISSUE THREAD

**Issue by PatrBal** [rickhowe/wrapwidth#7]
Hi,
thank you very much for this wonderful plugin. I have been waiting for this feature in Neovim for years!
Especially for my LaTeX writings.
My problem is that the virtual text feature is already used by `coc.nvim` which uses it to display "inlay Hints" provided by the `texlab` language server. This is illustrated by the following screenshots.

1) Original text in Neovim with neither "inlay Hints" nor "WrapWidth" enabled:
<img width="1456" alt="Screenshot 2024-01-22 at 16 33 15" src="https://github.com/rickhowe/wrapwidth/assets/44707588/d6170480-7cf4-470d-bcef-045e00ee57d9">
2) Same text in Neovim with `texlab`  "inlay Hints" enabled but with "WrapWidth" disabled:
<img width="1458" alt="Screenshot 2024-01-22 at 16 38 32" src="https://github.com/rickhowe/wrapwidth/assets/44707588/0d721149-f2cd-4b9f-9f99-d149a62fa3b6">
3) Original text with `texlab`  "inlay Hints" disabled and with "WrapWidth" enabled:
<img width="1453" alt="Screenshot 2024-01-22 at 16 40 26" src="https://github.com/rickhowe/wrapwidth/assets/44707588/3f45f992-e952-42ee-b209-70bd310babd0">
4) Finally, the same text with both  `texlab`  "inlay Hints"  and  "WrapWidth" enabled:
<img width="1458" alt="Screenshot 2024-01-22 at 16 42 22" src="https://github.com/rickhowe/wrapwidth/assets/44707588/a01d1b8f-f5e2-4627-a2a7-9ac294e2cbaf">

So, there is clearly a problem of incompatibility of WrapWidth with other sources of virtual text. It is possible that there is no good solution to that problem (except having the functionnality hardcoded in the neovim core without any use of the virtual text feature). I don't know enough about the virtual text feature to answer that.

Do you think it is possible to have wrapwidth working when other sources of virtual text are enabled?

Thank you very much again, in any case, for the nice plugin.

Edit: FWIW this problem is discussed in the following issue https://github.com/neovim/neovim/issues/24804
But there may still remain an issue when the "virtual spaces" have to be inserted in the middle of a "inlay hint"...

**Comment by rickhowe** [comment]
In 3) and 4), it looks that the wrapwidth plugin insets virtual spaces at the same text position. That means wrapwidth, actually builtin `virtcol()` and `virtcol2col()` functions, does not count other virtual text. What happens if you apply wrapwidth after coc?  In 4), what happens if you disable wrapwidth (`:Wrapwidth 0`) and then enable it (`:Wrapwidth N`) again? If still fails, we would need to check how coc implements their virtual text.

In 1) and 2), `virtcol([1, '$'])` should be different, but if it is same, I have no idea.

FYI: Nvim 0.10.0 fixed a defect about virtual text last week. I would recommend to update your nvim.

**Comment by PatrBal** [comment]
Good catch! If I apply wrapwidth after inlay Hints, it works fine :
<img width="1455" alt="Screenshot 2024-01-23 at 10 48 32" src="https://github.com/rickhowe/wrapwidth/assets/44707588/6d19c4a2-f07f-4a1d-951b-7b83595c92bf">
There is a problem only when inlay Hint is enabled after wrapwidth. Do you think it could be solved by adding a triggering event in wrapwidth? 

Also, if inlay Hints is subsequently disabled, we get the following result.
<img width="1466" alt="Screenshot 2024-01-23 at 11 09 42" src="https://github.com/rickhowe/wrapwidth/assets/44707588/b7912317-c239-4a39-9b7b-d7b871bb3983">

**Comment by rickhowe** [comment]
The virtual spaces are left unchanged after inlay Hints is disabled. The plugin can not do anything since there is no event to be triggered in this case. 

A workaround is to disable and then enable wrapwidth using `CursorHold`. For example:
`autocmd! CursorHold * Wrapwidth 0 | Wrapwidth 78`

Another one is to intentionally fire events:
`autocmd! CursorHold * doautocmd wrapwidth BufHidden * | doautocmd wrapwidth WinEnter *`

You might want to decrease the value of `updatetime` option.

Anyway, you have to always apply wrapwidth after coc but I can not find a good solution now.

**Comment by PatrBal** [comment]
Good idea, thanks! I have tested both and both work.
In my opinion, the second one is better as it allows for changes of the wrap column, and I will go with this one.
It is not perfect though as it is no more possible to toggle wrapwidth with ":Wrapwidth 0".

**Comment by rickhowe** [comment]
FYI: [RedrawWrapwidth.vim](https://github.com/rickhowe/sample/blob/main/RedrawWrapwidth.vim)
CursorHold happens very often. I found another workaround using map. When you notice the problem, press Ctrl-L and redraw wrapwidth spaces as well as vim screen.

**Comment by PatrBal** [comment]
I agree that CursorHold happens very often and should be therefore avoided. I am currently digging into the User Events of coc.nvim to improve the whole thing.
Your [RedrawWrapwidth.vim](https://github.com/rickhowe/sample/blob/main/RedrawWrapwidth.vim) solves the problem that wrapwidth could no more be turned off with ":Wrapwidth 0". However, I believe that `s:FindOtherVT` should be removed. Indeed, in the last screenshot above, there are no other virtual text than the wrapwidth virtual spaces and we need to redraw wrapwidth spaces nonetheless.
Incidentally, your function s:FindOtherVT does not work properly on my Neovim: it always says that there is no other Virtual text, even in cases where there is.

**Comment by rickhowe** [comment]
>Indeed, in the last screenshot above, there are no other virtual text than the wrapwidth virtual spaces and we need to redraw wrapwidth spaces nonetheless.

I see. 
`nnoremap <silent> <C-L> :<C-U>doautocmd wrapwidth BufHidden<CR>:<C-U>doautocmd wrapwidth BufWinEnter<CR><C-L>`
should work in any cases.

>Incidentally, your function s:FindOtherVT does not work properly on my Neovim: it always says that there is no other Virtual text, even in cases where there is.

I will check. Probably I missed to find 'virt_lines' as well.

**Comment by rickhowe** [comment]
I revised [RedrawWrapwidth.vim](https://github.com/rickhowe/sample/blob/main/RedrawWrapwidth.vim). It now checks if each virtual space exists at the right edge and check if other virtual text exists in other than wrapwidth lines. Then redraws each line using `:Wrapwidth` commd. Hope this helps.
